pkg/controllers/cluster/cluster_controller.go

```diff
@@ -238,8 +239,13 @@ func (c *Controller) syncCluster(ctx context.Context, cluster *clusterv1alpha1.C
 	return c.ensureFinalizer(ctx, cluster)
 }
 
-func (c *Controller) removeCluster(ctx context.Context, cluster *clusterv1alpha1.Cluster) (controllerruntime.Result, error) {
+func (c *Controller) removeCluster(ctx context.Context, cluster *clusterv1alpha1.Cluster) (ret controllerruntime.Result, retErr error) {
 	// add terminating taint before cluster is deleted
+	defer func() {
+		if retErr == nil {
```

Comment:
The `retErr == nil` can not ensure the `Cluster` has been successfully deleted.
It's possible that the cluster is still waiting for the next retry attempt, see https://github.com/karmada-io/karmada/blob/b605d9d6371885cbf0d850eb734ae3bedb98d576/pkg/controllers/cluster/cluster_controller.go#L260-L261